[MRG+1] Fix DummyClassifier bug with putting arrays into lists.

nsorros opened the pull request:
#### What does this implement/fix? Explain your changes.
It fixes an error thrown by dummy classifier when passing a dataframe y as explained more in depth in the issue. I enhance the check for output_2d to test whether the 2nd dimension is 1 in which case I use `column_or_1d` to transform the output.

#### Any other comments?
This is my first PR so any help and comments are greatly appreciated.

Fixes #10786

nsorros committed:
Add column_or_1d to account for dataframe y

nsorros committed:
Add test

nsorros committed:
make the diff cleaner

jnothman commented:
Tests failing.
Pease state "fixes #xxx" in the description, not title, so that GitHub will link it and automatically close the issue when the PR is merged

nsorros committed:
switched pandas import for 2d array

nsorros commented:
@jnothman some tests, not all, were complaining about the pandas import. any idea why? i removed it and replaced the array a matrix representation of a dataframe which replicates what was creating the problem. fingers crossed 🤞

jnothman commented:
We don't require Pandas to run scikit-learn, nor do we have it installed in
all testing instances. Tests that require Pandas must be skipped if Pandas
cannot be imported. pytest.importorskip should help.

On 10 April 2018 at 18:47, Nick Sorros <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> some tests, not all, were
> complaining about the pandas import. any idea why? i removed it and
> replaced the array a matrix representation of a dataframe which replicates
> what was creating the problem. fingers crossed 🤞
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10926#issuecomment-380023377>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6_5vI9bW4FeB1sjJl6y7Vo5pYxqcks5tnHGlgaJpZM4TJr-K>
> .
>

nsorros commented:
@jnothman Can I get some thoughts on this PR? The pandas dependency is now removed so it should be good to go.

lesteve commented on sklearn/dummy.py:
```diff
@@ -122,7 +123,11 @@ def fit(self, X, y, sample_weight=None):
-        self.output_2d_ = y.ndim == 2
+        self.output_2d_ = (y.ndim == 2) and (y.shape[1] > 1)
+
+        if not self.output_2d_:
```
Rather than resorting to custom code, can you not use `check_X_y` as I mentioned in the associated issue?

nsorros replied:
As far as I understood it the problem with using `check_X_y` is that since multi_output needs to be true in this case, `check_X_y` will not transform y to be 1d. We can set the flag to false if the output is not 2d but this ends up as the same code as above. If I am missing something let me know.

lesteve replied:
I am not 100% sure (and don't have time to dig into it right now) but it feels like you are using custom validation code where `check_X_y` some  arguments may need to be set properly for this use case.

The advantage of using a `check_*` function is that it is thoroughly used and tested across the code base, fixes to `check_*` will benefit you as well. By using custom code it is quite easy to introduce bugs.

Maybe what would make it easier to review is if you could summarize why `check_X_y` and/or `check_array` is not possible. Bonus points if you can add snippets because that sometimes makes it clearer than words.

nsorros replied:
That is the snippet that demonstrates what I was saying. The problem is that since we accept multi_outputs then check_X_y does not reduce the dimension to 1d in the special case where the second dimension is just 1 thus a vector e.g. shape=(x,1).

`import numpy as np

from sklearn.utils.validation import check_X_y

X = np.array([[0], [0], [0], [0]])
y_1d = np.array([1, 2, 1, 1])
y_2d = np.array([[1], [2], [1], [1]])

output_2d_ = (y_2d.ndim == 2) 

X_new, y_new = check_X_y(X, y_2d, multi_output=output_2d_)
print(y_2d.ndim)
print(y_new.ndim)`

nsorros replied:
Same solution to the problem exists in other parts of the code, for example in multi_layer_perceptron. 
```
915 if y.ndim == 2 and y.shape[1] == 1:
916  y = column_or_1d(y, warn=True)
```

This bug is mainly fixed in other classifiers but persists in the dummy one. One alternative would be to add a function in `validation.py` that checks whether y is multi_output and replace the repeatable custom code with that to achieve the benefits of reusability you are describing.

lesteve replied:
OK thanks for the details! Maybe we can use `warn=True` in `column_or_1d` so that there is a warning similarly to what is done in other places?

Side-comment: to link to code inside github like you can use this [nice feature](https://help.github.com/articles/creating-a-permanent-link-to-a-code-snippet/).

nsorros replied:
👍 Makes sense. I am uploading a new version with `warn=True`. Anything else that needed to be done in order to be ready to merge?

lesteve commented on sklearn/tests/test_dummy.py:
```diff
@@ -101,6 +100,23 @@ def test_most_frequent_and_prior_strategy():
                                       clf.class_prior_.reshape((1, -1)) > 0.5)
 
 
+def test_most_frequent_and_prior_strategy_with_pandas_dataframe():
```
Good stuff this is a non-regression test indeed since this fails in master. I think you should rename it though since it does not use a pandas dataframe at all.

lesteve commented on sklearn/tests/test_dummy.py:
```diff
@@ -101,6 +100,23 @@ def test_most_frequent_and_prior_strategy():
 
 
+def test_most_frequent_and_prior_strategy_with_pandas_dataframe():
+    X = [[0], [0], [0], [0]]  # ignored
```
I am not sure what you mean by this comment (maybe that X is irrelevant for this issue). In any case I think just remove the comment.

lesteve commented on sklearn/tests/test_dummy.py:
```diff
@@ -101,6 +100,23 @@ def test_most_frequent_and_prior_strategy():
 
+def test_most_frequent_and_prior_strategy_with_pandas_dataframe():
+    X = [[0], [0], [0], [0]]  # ignored
+    y = [[1], [2], [1], [1]]
```
So you copied this test from the function above and just added brackets around y so that it is 2d. Maybe something simpler would be to compare the `predict` or `y_1d` and `y_2d`.

lesteve reviewed (commented):
Some comments

nsorros committed:
change test to a simple comparison between 1d and 2d y

nsorros committed:
flake8 errors in tests

nsorros committed:
Add warning to acieve consistent behaviour as in other classifiers

jnothman reviewed (commented):
I agree with not using check_X_y but for different reasons: we should avoid placing constraints on X.

I don't quite understand why we a column input means predictions should be 1d. Is that consistent with elsewhere in the library?

jnothman commented:
Please add an entry to the change log at `doc/whats_new/v0.20.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

nsorros committed:
Update whats new file

nsorros commented:
Added a what's new entry. Do I need to mention the code reviewers as well? Is there anything else I need to do before being ready to merge?

jnothman commented:
Sorry for the slow reply. Did you answer my question above before my what's new comment?

nsorros commented:
@jnothman could you elaborate a bit more on what you mean? the input is not necessarily 1 column.

amueller commented:
@jnothman can you remember what exactly the convention here is? I don't think we have that written down, except maybe in the common tests?

jnothman commented:
no, I'm not sure, and I'm not available to investigate

nsorros commented:
@jnothman You got it right. I confused input to mean X. So if input y is of shape (n_samples, 1) which is 2D it is reduced to (n_samples,) which is 1D.

@amueller not sure what the convention is but this happens in other parts of the codebase as well. One way this problem arises is by passing a pandas dataframe y because scikit understands that to be 2d. Other classifiers are solving the problem in a similar fashion (example earlier on in my comments).

amueller commented:
Also see #9169 :-/ I considered this a bug. I have to double check though, it's been a while.

jnothman commented:
We should make sure to fix this one way or another

adrinjalali commented on doc/whats_new/v0.20.rst:
```diff
@@ -402,6 +402,10 @@ Decomposition, manifold learning and clustering
   :class:`mixture.BayesianGaussianMixture`. :issue:`10740` by :user:`Erich
   Schubert <kno10>` and :user:`Guillaume Lemaitre <glemaitre>`.
 
+- Fixed a bug in :class:`dummy.DummyClassifier` where 1d dimensional y with
```
could you please move this entry to `v0.21.rst` @nsorros ?

adrinjalali commented:
Could you please also rebase on master @nsorros ?

adrinjalali commented on sklearn/dummy.py:
```diff
@@ -122,7 +123,11 @@ def fit(self, X, y, sample_weight=None):
+        self.output_2d_ = (y.ndim == 2) and (y.shape[1] > 1)
+
+        if not self.output_2d_:
+            y = column_or_1d(y, warn=True)
```
This is reversed right after when `y` is reshaped to `(-1, 1)`. At least as far as the added tests here go, This if and `column_or_1d` is not needed and the change for `output_2d_` fixes the issue. Shouldn't the same fix apply to dummyregressor?
